### ISSUE TITLE
deeplearning-setをTensorflow v2に変更

### ISSUE THREAD

**Issue by OsugiTakuya** [ex-hasuolab/class-activation-map#22]
Tensorflow v2 公式Tutorialメモ

チュートリアルは6つ
3つは初心者用、3つは上級者用

**＜初心者用＞**
1. kerasを利用してモデル構築、model.fit()によるMNIST学習。  
これまで見てきたkerasの使い方と同じ。
2. 1.のデータをファッションMNIST（服、靴等の画像）に変更したサンプル。  
1.とやっていることは変わらない  
3. CSVファイルからデータを読みこみ、前処理してmodel.fitで学習するサンプル。  
データ前処理のちょっと面倒なところが書いている。  
CSVデータは、tensorflow関数によりのdatasetインスタンスとして読み込まれる （iter()によりイテレータに変換可能なクラス）。  
**model.fit()の引数にdatasetインスタンスを直接入れて学習。**

**＜上級者用＞**
1. MNIST学習サンプル。モデルの作成はkerasと同じ。  
学習はmodel.fitを使わず、  
　　訓練データから予測値出す→損失関数求める→勾配求める→パラメータ更新  
の流れで学習する（@tf.functionでデコレートした関数内で実装可能）。  
2. ・tf.Tensorの使い方、tf.addやtf.reduce_sumの使い方。  
・GPUの使い方（実行するデバイスの指定）  
・tf.data.Datasetの基本的な使い方
3. 1台のマシン上の複数GPUでの計算方法  
ぱっと見た感じ「このデータをこのGPUに割り振る」などしなくても勝手に割りふってくれそう。  
**ここでさりげなくTensorBoard、モデルチェックポイント 、学習率スケジューラ のコールバックの説明がされる。  
model.save()も書いている（どの形式で保存されるかは書いてない）**

**まとめとコメント**
+ 準備するデータはnumpy.arrayでも良いし、tf.data.Datasetでも良い。  
datasetの方は、dataset.shuffle(30), dataset.batch(50), dataset.repeat(10)（繰り返し出力する回数）など設定可能だが、augmentationが出来るわけではなさそう。  
augmentationはlayers.experimental.preprocessing.Resizing()など、層に追加することでやっているサンプルがあった。  
+ モデルの作り方はkerasと全く同じ  
+ 学習はkerasのようにmodel.fitでも良いし、勾配を計算してパラメータ更新でも良い。

**Comment by OsugiTakuya** [comment]
sparse_softmax_cross_entropy関数について  
https://qiita.com/tk_01/items/27e155f58da135cf47c2

**Comment by OsugiTakuya** [comment]
「【初心者向け】Tensorflow Object Detection API を使ってみた」  
https://qiita.com/IchiLab/items/fd99bcd92670607f8f9b  

→Object Detection APIはTensorFlow2.0には対応していないっぽい（2020年6月時点）

**Comment by TakehikoEsaka** [comment]
# tensorflow2のチュートリアル
[リンク](https://www.tensorflow.org/tutorials?hl=ja)
※ 内容の相関は分かりにくいけど，読み込んだらそれなりの情報おちてそう
（tensorboard・tensorflow dataset・tensorflow hub）

# Classificationは簡単にtf2で使える

```
IMG_SHAPE = IMG_SIZE + (3,)
base_model = tf.keras.applications.MobileNetV2(input_shape=IMG_SHAPE,
                                               include_top=False,
                                               weights='imagenet')
```
[tutorial](https://www.tensorflow.org/tutorials/images/transfer_learning?hl=ja)
[api仕様](https://www.tensorflow.org/api_docs/python/tf/keras/applications)

# Objetc Detection APIがTF2に対応（2020/7/10）
[Github Repositry](https://github.com/tensorflow/models/tree/master/research/object_detection)
[使えるnetwok一覧はmodel zoo](https://github.com/tensorflow/models/blob/da23acba8ecb8c0e7c9a83cdb9f10092895c9dcc/research/object_detection/g3doc/tf2_detection_zoo.md)
[colab tutorialはここ](https://github.com/tensorflow/models/tree/master/research/object_detection/colab_tutorials)

[Blog](https://blog.tensorflow.org/2020/07/tensorflow-2-meets-object-detection-api.html)

参考になりそうな記事
１：[gitabリポジトリの中の構造について解説](https://www.koi.mashykom.com/tensorflow.html#fd02)
２：[Anotation ツールのしょうかい](https://horomary.hatenablog.com/entry/2018/11/26/214048)
３：[Object Detection APIの使い方](https://qiita.com/IchiLab/items/fd99bcd92670607f8f9b)
４：[tensorflow2 object detection APIの推論使い方](https://qiita.com/karaage0703/items/8c3197d11f61812546a9)
５：[tesnsorflow liteで動かす方法](https://qiita.com/PINTO/items/69790c2f5b0416b5ec2e)
６：[Object Detection ToolをGUIで操作するツールを作成した人がいる](https://qiita.com/karaage0703/items/8567cc192e151bac3e50)
７：[ObjectDection APIの問題点について](https://karaage.hatenadiary.jp/entry/2019/05/27/073000)

すごい人：
https://qiita.com/karaage0703：組込系の人
https://qiita.com/PINTO：組込系の人
https://qiita.com/Kazuhito：組込系の人

**Comment by OsugiTakuya** [comment]
# Object Detection API を利用して転移学習するには？
以下のサンプルがまさにこれをやってる  
「Eager Few Shot Object Detection Colab」  
https://github.com/tensorflow/models/blob/master/research/object_detection/colab_tutorials/eager_few_shot_od_training_tf2_colab.ipynb  

**＜モデル作成＞**  
このサンプルノートブック中の「Create model and restore weights for all but last layer」と書かれた箇所の2つめのセル  
```
<略>
detection_model = model_builder.build(
      model_config=model_config, is_training=True)
<略>
fake_model = tf.compat.v2.train.Checkpoint(
          _feature_extractor=detection_model._feature_extractor,
          _box_predictor=fake_box_predictor)
ckpt = tf.compat.v2.train.Checkpoint(model=fake_model)
ckpt.restore(checkpoint_path).expect_partial()
```
+ model_builderは`from object_detection.builders import model_builder`からのもの。  
**detection_modelはkerasのmodelに対応したものであり、`detection_model.predict(画像)`等が使える。**  
**しかし Object Detection API 特有のモデルであるため、datageneratorに対応しているかどうかは要調査。**
+ どこかからcheckpointを読み込んでdetection_modelに反映させている（っぽい）

**＜学習＞**
このサンプルノートブック中の「Eager mode custom training loop」と書かれているセル 
```
def get_model_train_step_function(model, optimizer, vars_to_fine_tune):

    <略>

    @tf.function
    def train_step_fn(image_tensors,
                    groundtruth_boxes_list,
                    groundtruth_classes_list):

        <略>

        with tf.GradientTape() as tape:
            preprocessed_images = tf.concat(
                      [detection_model.preprocess(image_tensor)[0]
                        for image_tensor in image_tensors], axis=0)
            prediction_dict = model.predict(preprocessed_images, shapes)
            losses_dict = model.loss(prediction_dict, shapes)
            total_loss = losses_dict['Loss/localization_loss'] + losses_dict['Loss/classification_loss']
            gradients = tape.gradient(total_loss, vars_to_fine_tune)
            optimizer.apply_gradients(zip(gradients, vars_to_fine_tune))  
        return total_loss
```
+ preprocessed_images に画像データのTensorが格納
+ model.predict で現在のパラメータ値での予測値を出力
**ここのmodelはkerasのモデルとは異なり、**
`from object_detection.builders import model_builder`
**のmodel_builderを使ってビルドされたもの。checkpointから重みを読みこんで反映させたものっぽい**

**Comment by TakehikoEsaka** [comment]
google-colab local Dockerfile

1 : https://github.com/googlecolab/backend-container/blob/bcbbf44/containers/Dockerfile
2 : https://github.com/ikeyasu/docker-colab-local/blob/master/Dockerfile

**Comment by OsugiTakuya** [comment]
# 9/5以降やること
Tensorflow 2.3.0で、ローカル環境でobject_detection APIを動かせる環境を作る  
+ protoc コマンドを動くようにする
+ Eagerサンプルが動くようにする
+ Dockerfile作る
